fix: Lambda cold-start timeout causes 503 on fresh deploy (validate-endpoints fails)

## Summary

The CI deploy job fails at the **Validate backend and frontend endpoints** step with a 503 immediately after a fresh `cdk deploy`. The root cause is the `BackendLambda` timing out during its cold start, making the function unreachable for the endpoint health check.

Observed in run [25608114650 / job 75173393981](https://github.com/leonarduk/allotmint/actions/runs/25608114650/job/75173393981).

## Symptoms

CI step 21 (`Validate backend and frontend endpoints`):
```
curl: (22) The requested URL returned error: 503
##[error]Process completed with exit code 22.
```

AWS Lambda logs (from CloudWatch):
```
INIT_REPORT Init Duration: 10010.59 ms  Phase: init   Status: timeout
INIT_REPORT Init Duration: 30001.00 ms  Phase: invoke Status: timeout
```

Also logged during init:
```
[ERROR] Failed to fetch price snapshot prices/latest_prices.json from bucket … NoSuchKey
[ERROR] Failed to load alert thresholds from S3 … NoSuchKey
[ERROR] Failed to load push subscriptions from S3 … NoSuchKey
```

## Root causes

### 1 — 128 MB Lambda memory (CDK default)
`BackendLambda` has no `memory_size` set, so it defaults to 128 MB. Lambda allocates vCPU proportionally to memory, meaning the function runs with minimal CPU. Importing the heavy Python dependencies (pandas, numpy, pyarrow, selenium, …) under these constraints takes long enough to blow past the init-phase budget.

### 2 — Module-level S3 calls in `backend/alerts.py`
`alerts.py` called `_load_settings()` and `_load_subscriptions()` at **module import time** (lines 221–222). On a fresh deployment the S3 bucket contains no alert data, so each call fires an S3 `GetObject` that fails with `NoSuchKey`. These synchronous network round-trips run during the Lambda init phase and compound the timeout.

### 3 — Empty-dict retry loop
The load guard `if _USER_THRESHOLDS:` is falsy for an empty `{}`. When S3 returns no alert data the cache ends up as `{}`, the guard never short-circuits, and **every subsequent warm-Lambda request** retries S3 unnecessarily.

## Fix

See PR #&lt;TBD&gt; (`fix/lambda-cold-start-timeout`):

| File | Change |
|---|---|
| `cdk/stacks/backend_lambda_stack.py` | `memory_size=512` (MiB) on `BackendLambda` |
| `backend/alerts.py` | Remove module-level `_load_settings()` / `_load_subscriptions()` calls; replace empty-dict sentinel with `_SETTINGS_LOADED` / `_SUBSCRIPTIONS_LOADED` boolean flags |
| `tests/test_alerts_storage.py` | Reset new sentinel flags in `clear_caches` fixture |

## Acceptance criteria

- [ ] CI `Validate backend and frontend endpoints` step passes on a fresh `cdk deploy`
- [ ] Lambda cold-start completes within the 30 s function timeout
- [ ] No repeated S3 retries when alert bucket keys are absent

--- a/backend/alerts.py
+++ b/backend/alerts.py
@@ -109,7 +109,6 @@ def _load_settings() -> None:
     global _USER_THRESHOLDS, _SETTINGS_LOADED
     if _SETTINGS_LOADED or _USER_THRESHOLDS:
         return
-    _SETTINGS_LOADED = True
     bucket = _data_bucket()
     if bucket:
         s3 = _s3_client()
@@ -119,20 +118,33 @@ def _load_settings() -> None:
                 data = json.loads(obj["Body"].read().decode())
                 if isinstance(data, dict):
                     _USER_THRESHOLDS = _parse_thresholds(data)
+                _SETTINGS_LOADED = True
                 return
-            except Exception:
-                logging.getLogger("alerts").exception("Failed to load alert thresholds from S3")
+            except Exception as exc:
+                code = None
+                try:
+                    code = exc.response['Error']['Code']
+                except (AttributeError, KeyError, TypeError):
+                    pass
+                if code == 'NoSuchKey':
+                    _SETTINGS_LOADED = True
+                else:
+                    logging.getLogger("alerts").exception("Failed to load alert thresholds from S3")
+                return
     try:
         data = _SETTINGS_STORAGE.load()
     except Exception as exc:  # pragma: no cover - storage backend failures
         logger.warning('Failed to load user thresholds: %s', exc)
         _USER_THRESHOLDS = {}
+        _SETTINGS_LOADED = True
         return
     if not isinstance(data, dict):
         logger.warning('User thresholds data malformed: %r', data)
         _USER_THRESHOLDS = {}
+        _SETTINGS_LOADED = True
         return
     _USER_THRESHOLDS = _parse_thresholds(data)
+    _SETTINGS_LOADED = True
 
 
 def _load_subscriptions() -> None:
@@ -140,7 +152,6 @@ def _load_subscriptions() -> None:
     global _PUSH_SUBSCRIPTIONS, _SUBSCRIPTIONS_LOADED
     if _SUBSCRIPTIONS_LOADED or _PUSH_SUBSCRIPTIONS:
         return
-    _SUBSCRIPTIONS_LOADED = True
     bucket = _data_bucket()
     if bucket:
         s3 = _s3_client()
@@ -150,20 +161,33 @@ def _load_subscriptions() -> None:
                 data = json.loads(obj["Body"].read().decode())
                 if isinstance(data, dict):
                     _PUSH_SUBSCRIPTIONS = _parse_subscriptions(data)
+                _SUBSCRIPTIONS_LOADED = True
                 return
-            except Exception:
-                logging.getLogger("alerts").exception("Failed to load push subscriptions from S3")
+            except Exception as exc:
+                code = None
+                try:
+                    code = exc.response['Error']['Code']
+                except (AttributeError, KeyError, TypeError):
+                    pass
+                if code == 'NoSuchKey':
+                    _SUBSCRIPTIONS_LOADED = True
+                else:
+                    logging.getLogger("alerts").exception("Failed to load push subscriptions from S3")
+                return
     try:
         data = _SUBSCRIPTIONS_STORAGE.load()
     except Exception as exc:  # pragma: no cover - storage backend failures
         logger.warning('Failed to load push subscriptions: %s', exc)
         _PUSH_SUBSCRIPTIONS = {}
+        _SUBSCRIPTIONS_LOADED = True
         return
     if not isinstance(data, dict):
         logger.warning('Push subscriptions data malformed: %r', data)
         _PUSH_SUBSCRIPTIONS = {}
+        _SUBSCRIPTIONS_LOADED = True
         return
     _PUSH_SUBSCRIPTIONS = _parse_subscriptions(data)
+    _SUBSCRIPTIONS_LOADED = True
 
 def _save_settings() -> None:
     """Persist in-memory settings to configured storage."""

--- a/backend/alerts.py
+++ b/backend/alerts.py
@@ -121,16 +121,11 @@ def _load_settings() -> None:
                 _SETTINGS_LOADED = True
                 return
             except Exception as exc:
-                code = None
-                try:
-                    code = exc.response['Error']['Code']
-                except (AttributeError, KeyError, TypeError):
-                    pass
-                if code == 'NoSuchKey':
+                _resp = getattr(exc, 'response', {}) or {}
+                if _resp.get('Error', {}).get('Code') == 'NoSuchKey':
                     _SETTINGS_LOADED = True
-                else:
-                    logging.getLogger("alerts").exception("Failed to load alert thresholds from S3")
-                return
+                    return
+                logging.getLogger("alerts").exception("Failed to load alert thresholds from S3")
     try:
         data = _SETTINGS_STORAGE.load()
     except Exception as exc:  # pragma: no cover - storage backend failures
@@ -164,16 +159,11 @@ def _load_subscriptions() -> None:
                 _SUBSCRIPTIONS_LOADED = True
                 return
             except Exception as exc:
-                code = None
-                try:
-                    code = exc.response['Error']['Code']
-                except (AttributeError, KeyError, TypeError):
-                    pass
-                if code == 'NoSuchKey':
+                _resp = getattr(exc, 'response', {}) or {}
+                if _resp.get('Error', {}).get('Code') == 'NoSuchKey':
                     _SUBSCRIPTIONS_LOADED = True
-                else:
-                    logging.getLogger("alerts").exception("Failed to load push subscriptions from S3")
-                return
+                    return
+                logging.getLogger("alerts").exception("Failed to load push subscriptions from S3")
     try:
         data = _SUBSCRIPTIONS_STORAGE.load()
     except Exception as exc:  # pragma: no cover - storage backend failures

--- a/backend/alerts.py
+++ b/backend/alerts.py
@@ -104,6 +104,21 @@ def _parse_subscriptions(data: Dict) -> Dict[str, Dict]:
     return valid
 
 
+def _s3_error_code(exc: Exception) -> Optional[str]:
+    """Return an AWS-style error code from ``exc`` when one is present."""
+    response = getattr(exc, "response", {}) or {}
+    error = response.get("Error", {})
+    if not isinstance(error, dict):
+        return None
+    code = error.get("Code")
+    return code if isinstance(code, str) else None
+
+
+def ensure_alert_settings_loaded() -> None:
+    """Ensure the in-memory threshold cache has attempted its lazy load."""
+    _load_settings()
+
+
 def _load_settings() -> None:
     """Load threshold settings into memory from configured storage."""
     global _USER_THRESHOLDS, _SETTINGS_LOADED
@@ -121,25 +136,29 @@ def _load_settings() -> None:
                 _SETTINGS_LOADED = True
                 return
             except Exception as exc:
-                _resp = getattr(exc, 'response', {}) or {}
-                if _resp.get('Error', {}).get('Code') == 'NoSuchKey':
+                if _s3_error_code(exc) == "NoSuchKey":
                     _SETTINGS_LOADED = True
                     return
-                logging.getLogger("alerts").exception("Failed to load alert thresholds from S3")
+                logger.exception("Failed to load alert thresholds from S3")
     try:
         data = _SETTINGS_STORAGE.load()
     except Exception as exc:  # pragma: no cover - storage backend failures
-        logger.warning('Failed to load user thresholds: %s', exc)
+        logger.warning("Failed to load user thresholds: %s", exc)
         _USER_THRESHOLDS = {}
         _SETTINGS_LOADED = True
         return
     if not isinstance(data, dict):
-        logger.warning('User thresholds data malformed: %r', data)
+        logger.warning("User thresholds data malformed: %r", data)
         _USER_THRESHOLDS = {}
         _SETTINGS_LOADED = True
         return
     _USER_THRESHOLDS = _parse_thresholds(data)
     _SETTINGS_LOADED = True
+
+
+def ensure_push_subscriptions_loaded() -> None:
+    """Ensure the in-memory push subscription cache has attempted its lazy load."""
+    _load_subscriptions()
 
 
 def _load_subscriptions() -> None:
@@ -159,25 +178,25 @@ def _load_subscriptions() -> None:
                 _SUBSCRIPTIONS_LOADED = True
                 return
             except Exception as exc:
-                _resp = getattr(exc, 'response', {}) or {}
-                if _resp.get('Error', {}).get('Code') == 'NoSuchKey':
+                if _s3_error_code(exc) == "NoSuchKey":
                     _SUBSCRIPTIONS_LOADED = True
                     return
-                logging.getLogger("alerts").exception("Failed to load push subscriptions from S3")
+                logger.exception("Failed to load push subscriptions from S3")
     try:
         data = _SUBSCRIPTIONS_STORAGE.load()
     except Exception as exc:  # pragma: no cover - storage backend failures
-        logger.warning('Failed to load push subscriptions: %s', exc)
+        logger.warning("Failed to load push subscriptions: %s", exc)
         _PUSH_SUBSCRIPTIONS = {}
         _SUBSCRIPTIONS_LOADED = True
         return
     if not isinstance(data, dict):
-        logger.warning('Push subscriptions data malformed: %r', data)
+        logger.warning("Push subscriptions data malformed: %r", data)
         _PUSH_SUBSCRIPTIONS = {}
         _SUBSCRIPTIONS_LOADED = True
         return
     _PUSH_SUBSCRIPTIONS = _parse_subscriptions(data)
     _SUBSCRIPTIONS_LOADED = True
+
 
 def _save_settings() -> None:
     """Persist in-memory settings to configured storage."""
@@ -205,7 +224,7 @@ def _save_settings() -> None:
         _SETTINGS_STORAGE.save(_USER_THRESHOLDS)
     except Exception as exc:  # pragma: no cover - storage backend failures
         # Persistence failure should not block alerting
-        logger.error('Failed to save user thresholds to persistent storage: %s', exc)
+        logger.error("Failed to save user thresholds to persistent storage: %s", exc)
 
 
 def _save_subscriptions() -> None:
@@ -233,7 +252,7 @@ def _save_subscriptions() -> None:
     try:
         _SUBSCRIPTIONS_STORAGE.save(_PUSH_SUBSCRIPTIONS)
     except Exception as exc:  # pragma: no cover - storage backend failures
-        logger.error('Failed to save push subscriptions to persistent storage: %s', exc)
+        logger.error("Failed to save push subscriptions to persistent storage: %s", exc)
 
 
 def get_user_threshold(user: str, default: float = DEFAULT_THRESHOLD_PCT) -> float:

--- a/backend/alerts.py
+++ b/backend/alerts.py
@@ -56,9 +56,11 @@ except Exception as exc:  # pragma: no cover - configuration errors
 
 # In-memory cache of settings
 _USER_THRESHOLDS: Dict[str, float] = {}
+_SETTINGS_LOADED: bool = False
 
 # In-memory cache of push subscriptions
 _PUSH_SUBSCRIPTIONS: Dict[str, Dict] = {}
+_SUBSCRIPTIONS_LOADED: bool = False
 
 
 def _s3_client():
@@ -104,9 +106,10 @@ def _parse_subscriptions(data: Dict) -> Dict[str, Dict]:
 
 def _load_settings() -> None:
     """Load threshold settings into memory from configured storage."""
-    global _USER_THRESHOLDS
-    if _USER_THRESHOLDS:
+    global _USER_THRESHOLDS, _SETTINGS_LOADED
+    if _SETTINGS_LOADED or _USER_THRESHOLDS:
         return
+    _SETTINGS_LOADED = True
     bucket = _data_bucket()
     if bucket:
         s3 = _s3_client()
@@ -134,9 +137,10 @@ def _load_settings() -> None:
 
 def _load_subscriptions() -> None:
     """Load push subscription data into memory from configured storage."""
-    global _PUSH_SUBSCRIPTIONS
-    if _PUSH_SUBSCRIPTIONS:
+    global _PUSH_SUBSCRIPTIONS, _SUBSCRIPTIONS_LOADED
+    if _SUBSCRIPTIONS_LOADED or _PUSH_SUBSCRIPTIONS:
         return
+    _SUBSCRIPTIONS_LOADED = True
     bucket = _data_bucket()
     if bucket:
         s3 = _s3_client()
@@ -216,10 +220,6 @@ def _save_subscriptions() -> None:
         _SUBSCRIPTIONS_STORAGE.save(_PUSH_SUBSCRIPTIONS)
     except Exception as exc:  # pragma: no cover - storage backend failures
         logger.error('Failed to save push subscriptions to persistent storage: %s', exc)
-
-
-_load_settings()
-_load_subscriptions()
 
 
 def get_user_threshold(user: str, default: float = DEFAULT_THRESHOLD_PCT) -> float:

--- a/backend/quests/trail.py
+++ b/backend/quests/trail.py
@@ -232,6 +232,7 @@ def _normalise_threshold(raw: object) -> float | None:
 
 
 def _has_custom_threshold(user: str) -> bool:
+    alerts._load_settings()
     thresholds = getattr(alerts, "_USER_THRESHOLDS", {})
     if not isinstance(thresholds, dict):
         return False

--- a/backend/quests/trail.py
+++ b/backend/quests/trail.py
@@ -31,9 +31,7 @@ ONCE_XP_REWARD = 25
 # Storage helpers
 # ---------------------------------------------------------------------------
 
-_DEFAULT_TRAIL_URI = (
-    f"file://{(config.repo_root or Path(__file__).resolve().parents[1]) / 'data' / 'trail.json'}"
-)
+_DEFAULT_TRAIL_URI = f"file://{(config.repo_root or Path(__file__).resolve().parents[1]) / 'data' / 'trail.json'}"
 _TRAIL_STORAGE = get_storage(os.getenv("TRAIL_URI", _DEFAULT_TRAIL_URI))
 
 # Data format per user::
@@ -146,8 +144,7 @@ def _build_allowance_tasks(owners: Iterable[str]) -> List[TaskDefinition]:
             label_raw = account.replace("_", " ")
             account_label = label_raw.upper() if label_raw.upper() == "ISA" else label_raw.title()
             commentary = (
-                f"{account_label} allowance remaining: {_format_currency(remaining)}"
-                f" of {_format_currency(limit)}."
+                f"{account_label} allowance remaining: {_format_currency(remaining)}" f" of {_format_currency(limit)}."
             )
             tasks.append(
                 TaskDefinition(
@@ -232,7 +229,7 @@ def _normalise_threshold(raw: object) -> float | None:
 
 
 def _has_custom_threshold(user: str) -> bool:
-    alerts._load_settings()
+    alerts.ensure_alert_settings_loaded()
     thresholds = getattr(alerts, "_USER_THRESHOLDS", {})
     if not isinstance(thresholds, dict):
         return False
@@ -252,10 +249,7 @@ def _has_custom_threshold(user: str) -> bool:
         if default_threshold >= 1:
             default_variants.add(default_threshold / 100)
 
-    if any(
-        math.isclose(candidate, default, rel_tol=1e-9, abs_tol=1e-9)
-        for default in default_variants
-    ):
+    if any(math.isclose(candidate, default, rel_tol=1e-9, abs_tol=1e-9) for default in default_variants):
         return False
 
     return True
@@ -274,7 +268,7 @@ _AUTO_ONCE_BLOCKLIST = {
 def _sync_once_completion(user_data: Dict, task_id: str, should_complete: bool) -> None:
     """Ensure ``task_id`` reflects ``should_complete`` within ``user_data``."""
 
-    once_list = user_data.setdefault("once", [])
+    _ = user_data.setdefault("once", [])  # Keep manual completions initialized.
     auto_list = user_data.setdefault(_AUTO_ONCE_KEY, [])
 
     if should_complete:
@@ -475,6 +469,7 @@ def _save() -> None:
 # Public API
 # ---------------------------------------------------------------------------
 
+
 def get_tasks(user: str) -> Dict:
     """Return tasks and completion state for ``user``."""
     _load()
@@ -486,9 +481,7 @@ def get_tasks(user: str) -> Dict:
     daily_task_ids = [task.id for task in task_defs if task.type == "daily"]
     once_task_ids = [task.id for task in task_defs if task.type == "once"]
 
-    daily_completed = {
-        task_id for task_id in user_data["daily"].get(today, []) if task_id in daily_task_ids
-    }
+    daily_completed = {task_id for task_id in user_data["daily"].get(today, []) if task_id in daily_task_ids}
     user_data["daily"][today] = sorted(daily_completed)
     manual_once = [task_id for task_id in user_data.get("once", []) if task_id in once_task_ids]
     auto_once = [task_id for task_id in user_data.get(_AUTO_ONCE_KEY, []) if task_id in once_task_ids]
@@ -497,9 +490,7 @@ def get_tasks(user: str) -> Dict:
 
     tasks: List[Dict[str, object]] = []
     for task in task_defs:
-        completed = (
-            task.id in once_completed if task.type == "once" else task.id in daily_completed
-        )
+        completed = task.id in once_completed if task.type == "once" else task.id in daily_completed
         if task.id == "set_alert_threshold" and has_custom_threshold:
             completed = True
         tasks.append(

--- a/cdk/stacks/backend_lambda_stack.py
+++ b/cdk/stacks/backend_lambda_stack.py
@@ -238,6 +238,7 @@ class BackendLambdaStack(Stack):
             environment=backend_env,
             log_group=backend_log_group,
             timeout=Duration.seconds(30),
+            memory_size=512,
         )
         backend_fn.add_environment("APP_ENV", env)
 

--- a/tests/test_alerts_storage.py
+++ b/tests/test_alerts_storage.py
@@ -1,9 +1,16 @@
 import json
 import sys
 import types
+
 import pytest
 
 import backend.alerts as alerts
+
+
+class FakeS3Error(Exception):
+    def __init__(self, code):
+        super().__init__(code)
+        self.response = {"Error": {"Code": code}}
 
 
 @pytest.fixture(autouse=True)
@@ -62,6 +69,55 @@ def test_load_settings_falls_back_to_local_on_s3_error(monkeypatch):
     monkeypatch.setattr(alerts, "_SETTINGS_STORAGE", types.SimpleNamespace(load=loader))
 
     alerts._load_settings()
+    assert alerts._USER_THRESHOLDS == {"a": 0.4}
+
+
+def test_load_settings_does_not_retry_missing_s3_key(monkeypatch):
+    calls = []
+
+    class MissingKeyS3:
+        def get_object(self, Bucket, Key):
+            calls.append((Bucket, Key))
+            raise FakeS3Error("NoSuchKey")
+
+    def boom():
+        raise AssertionError("local load should not be used for missing S3 key")
+
+    monkeypatch.setattr(alerts, "_data_bucket", lambda: "bucket")
+    monkeypatch.setattr(alerts, "_s3_client", lambda: MissingKeyS3())
+    monkeypatch.setattr(alerts, "_SETTINGS_STORAGE", types.SimpleNamespace(load=boom))
+
+    alerts._load_settings()
+    alerts._load_settings()
+
+    assert calls == [("bucket", alerts._THRESHOLDS_KEY)]
+    assert alerts._SETTINGS_LOADED is True
+    assert alerts._USER_THRESHOLDS == {}
+
+
+def test_load_settings_s3_error_falls_back_to_local_once(monkeypatch):
+    s3_calls = []
+    local_calls = []
+
+    class BrokenS3:
+        def get_object(self, Bucket, Key):
+            s3_calls.append((Bucket, Key))
+            raise FakeS3Error("AccessDenied")
+
+    def loader():
+        local_calls.append("load")
+        return {"a": "0.4"}
+
+    monkeypatch.setattr(alerts, "_data_bucket", lambda: "bucket")
+    monkeypatch.setattr(alerts, "_s3_client", lambda: BrokenS3())
+    monkeypatch.setattr(alerts, "_SETTINGS_STORAGE", types.SimpleNamespace(load=loader))
+
+    alerts._load_settings()
+    alerts._load_settings()
+
+    assert s3_calls == [("bucket", alerts._THRESHOLDS_KEY)]
+    assert local_calls == ["load"]
+    assert alerts._SETTINGS_LOADED is True
     assert alerts._USER_THRESHOLDS == {"a": 0.4}
 
 
@@ -161,6 +217,55 @@ def test_load_subscriptions_falls_back_to_local_on_s3_error(monkeypatch):
     monkeypatch.setattr(alerts, "_SUBSCRIPTIONS_STORAGE", types.SimpleNamespace(load=loader))
 
     alerts._load_subscriptions()
+    assert alerts._PUSH_SUBSCRIPTIONS == {"u1": {"a": 1}}
+
+
+def test_load_subscriptions_does_not_retry_missing_s3_key(monkeypatch):
+    calls = []
+
+    class MissingKeyS3:
+        def get_object(self, Bucket, Key):
+            calls.append((Bucket, Key))
+            raise FakeS3Error("NoSuchKey")
+
+    def boom():
+        raise AssertionError("local load should not be used for missing S3 key")
+
+    monkeypatch.setattr(alerts, "_data_bucket", lambda: "bucket")
+    monkeypatch.setattr(alerts, "_s3_client", lambda: MissingKeyS3())
+    monkeypatch.setattr(alerts, "_SUBSCRIPTIONS_STORAGE", types.SimpleNamespace(load=boom))
+
+    alerts._load_subscriptions()
+    alerts._load_subscriptions()
+
+    assert calls == [("bucket", alerts._SUBSCRIPTIONS_KEY)]
+    assert alerts._SUBSCRIPTIONS_LOADED is True
+    assert alerts._PUSH_SUBSCRIPTIONS == {}
+
+
+def test_load_subscriptions_s3_error_falls_back_to_local_once(monkeypatch):
+    s3_calls = []
+    local_calls = []
+
+    class BrokenS3:
+        def get_object(self, Bucket, Key):
+            s3_calls.append((Bucket, Key))
+            raise FakeS3Error("AccessDenied")
+
+    def loader():
+        local_calls.append("load")
+        return {"u1": {"a": 1}, "u2": "bad"}
+
+    monkeypatch.setattr(alerts, "_data_bucket", lambda: "bucket")
+    monkeypatch.setattr(alerts, "_s3_client", lambda: BrokenS3())
+    monkeypatch.setattr(alerts, "_SUBSCRIPTIONS_STORAGE", types.SimpleNamespace(load=loader))
+
+    alerts._load_subscriptions()
+    alerts._load_subscriptions()
+
+    assert s3_calls == [("bucket", alerts._SUBSCRIPTIONS_KEY)]
+    assert local_calls == ["load"]
+    assert alerts._SUBSCRIPTIONS_LOADED is True
     assert alerts._PUSH_SUBSCRIPTIONS == {"u1": {"a": 1}}
 
 

--- a/tests/test_alerts_storage.py
+++ b/tests/test_alerts_storage.py
@@ -9,7 +9,9 @@ import backend.alerts as alerts
 @pytest.fixture(autouse=True)
 def clear_caches(monkeypatch):
     monkeypatch.setattr(alerts, "_USER_THRESHOLDS", {})
+    monkeypatch.setattr(alerts, "_SETTINGS_LOADED", False)
     monkeypatch.setattr(alerts, "_PUSH_SUBSCRIPTIONS", {})
+    monkeypatch.setattr(alerts, "_SUBSCRIPTIONS_LOADED", False)
 
 
 def test_parse_thresholds_parses_valid_entries():
@@ -329,7 +331,9 @@ def test_threshold_and_subscription_persistence(monkeypatch):
 
     # Clear caches and reload to verify persistence via getters
     alerts._USER_THRESHOLDS = {}
+    alerts._SETTINGS_LOADED = False
     alerts._PUSH_SUBSCRIPTIONS = {}
+    alerts._SUBSCRIPTIONS_LOADED = False
 
     assert alerts.get_user_threshold("u1") == 0.9
     assert alerts.get_user_push_subscription("u1") == {"x": 1}

